### PR TITLE
Add i3/Sway support

### PR DESCRIPTION
**Pull request by nydragon**
This pr adds support for retrieving data from Sway's IPC.

Features:

- [x] Get last IPC object
- [x] Dispatch Sway commands
- [x] Workspaces
   - [x] Retrieve all
   - [x] Listen to data changes via signals
   - [x] Force full refresh of workspaces
   - [x] Get active 
   - [x] Live update
- [x] Monitors
   - [x] Retrieve all
   - [x] Listen to data changes via signals
   - [x] Force full refresh of monitors
   - [x] Get active
   - [x] Live update
   
Additional Todo:
- [x] documentation

### CODE REVIEW

**Comment by outfoxxed** [comment]
Looks largely copied from hyprland ipc at this point so I figured I'd mention a few things:
- Make sure you only copy what makes sense to copy, don't keep what doesn't fit.
- If you can leave the dispatch socket open you should. You can't in hyprland, idk about i3/sway
- If you aren't implementing any sway specific features, this should work with i3 and be named as such.

**Comment by nydragon** [comment]
Thanks for the feedback, I wanted to stay mainly in line with the existing code, but I do have some questions:

1. Sway calls monitors "outputs", should that be reflected in Quickshell or is it better to be have uniform naming across different compositors ? Or "dispatch" is called "run commands"
2. Should I move the x11 and wayland folders into a compositor folder and create a `shared` folder in there? Or where would I put the sway-i3 code ?
3. We could abstract a part of the logic, namely the workspace/monitors refresh functions, across Hyprland & Sway/i3, does that make sense ?

**Comment by outfoxxed** [comment]
1. Monitor. Dispatch under hyprland is only for commands with no response as we don't currently have a promise type.
2. For now src/x11/i3/ipc is fine. Hyprland is under the wayland folder because it has to use cmake functions defined in the wayland folder.
3. Yes. This should happen, but not in this PR.

Also make sure you have a way to turn a sway monitor into a qs monitor.

**Comment by outfoxxed** [comment]
You've force pushed a lot, this ready for review?

**Comment by nydragon** [comment]
Yep, you could review it,

I would still want to implement more i3/Sway IPC features, but I think that will be for a later PR

File: CMakeLists.txt
Comment (outfoxxed) [inline CMakeLists.txt:59]: Add a note in BUILD.md

File: src/x11/i3/ipc/CMakeLists.txt
Comment (outfoxxed) [inline src/x11/i3/ipc/CMakeLists.txt]: Remove Quickshell from DEPENDENCIES. add_module_deps_light covers that.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: lowercase

File: src/x11/i3/ipc/connection.hpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.hpp]: do we need 2?

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: Consider using std::bit_cast or QDataStream

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: grammar

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: qWarning -> qCWarning everywhere

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: qinfo -> qcinfo and this looks debug worthy at most

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: debug

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: same as above

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: why mList? `m` conventionally means "member".

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: use all == unless i'm missing something

File: src/x11/i3/ipc/connection.hpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.hpp]: what lint is failing?

File: src/x11/i3/ipc/qml.hpp
Comment (outfoxxed) [inline src/x11/i3/ipc/qml.hpp]: link https://i3wm.org/docs/userguide.html#list_of_commands instead probably

**Review by outfoxxed — CHANGES_REQUESTED** [review]
Bunch of nitpicks. I assume you've gone through and tested everything manually as well, run the lints, etc. If you have a tester file send that along as well and I'll give it a try before merging.

File: src/x11/i3/ipc/connection.cpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.cpp]: Thought it'd be more concise than matching all 8 enum members

File: src/x11/i3/ipc/connection.hpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.hpp]: It doesn't like screaming snake case

File: src/x11/i3/ipc/connection.hpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.hpp]: Didn't find a way to circumvent a race condition coming from the live ipc event sending a message while still waiting for the command response, since I am returning the reply to the QS user

**Comment by nydragon** [comment]
I have just noticed a crash related to those changes, don't know how to replicate it yet but I would need to fix that before this PR can progress

edit: fixed, was a lot more straightforward than I thought

**Comment by outfoxxed** [comment]
Consider using QDataStream transactions instead. As is, if you break halfway through an event then the following event will be invalid and so on. check ipc.cpp for an example

File: src/x11/i3/ipc/connection.hpp
Comment (outfoxxed) [reply src/x11/i3/ipc/connection.hpp]: ah yeah. use UpperCamelCase with an enum class.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [reply src/x11/i3/ipc/connection.cpp]: this doesn't look like it matches all the enum members as is

File: src/x11/i3/ipc/connection.hpp
Comment (outfoxxed) [reply src/x11/i3/ipc/connection.hpp]: If you're returning all events and not just requested ones you can use a single socket since you'd be racing anyway. Otherwise 2 is fine

File: src/x11/i3/ipc/connection.cpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.cpp]: It should be matching all except for the Unknown member

File: src/x11/i3/ipc/connection.hpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.hpp]: I am emitting all events through rawEvent, but dispatch returns to the user the IPC's reply to the requested command

**Comment by nydragon** [comment]
> Consider using QDataStream transactions instead. As is, if you break halfway through an event then the following event will be invalid and so on. check ipc.cpp for an example

I ended up doing that with QByteArrays instead, if I encounter a broken message i try to find the next occurrence of the magic sequence and skip to there. However this is under the assumption that the Sway IPC does not write partial messages to the socket, and it does not seem to do so.

**Comment by outfoxxed** [comment]
Do we ever encounter broken messages? i3 shouldn't be sending those at all. I think the best solution to read the header would be a QDataStream transaction, which lets you do a partial read pretty easily.

Basically:
```cpp
stream.startTransaction()
stream.startTransaction()
stream >> magic >> header bits >> size;
if (!stream.commitTransaction()) return;
stream >> data; // assuming you need size to find data. nested commitTransaction works as a barrier for the integrity of values read previosuly
if (!stream.commitTransaction()) return;
```
You can nest transactions and it works pretty well. This should be more reliable than skipping events. Failing commits roll back the stream, and if there's enough data on the next readyread it should work.

**Comment by nydragon** [comment]
Alrighty, that should be implemented now, here is also a file to test the IPC impl:

<details><summary>Details</summary>
<p>

```qml
import QtQuick
import Quickshell
import Quickshell.I3

ShellRoot {
    PanelWindow {
        id: panel
        Timer {
            id: timer
        }
        Component.onCompleted: {
            const program = "foot"; // change to an installed program
            const longLog = false; // toggle logging of workspace & monitor objects
            const stringify = v => JSON.stringify(v, null, 2);
            console.log("Connecting to", I3.socketPath);
            I3.rawEvent.connect(e => {
                console.log("Event:", e.type);
            });
            I3.focusedWorkspaceChanged.connect(() => {
                console.log("Changed to workspace", I3.focusedWorkspace.name, I3.focusedMonitor.focusedWorkspace.name);
                console.log("Focused monitor's focused workspace and actual focused workspace should be the same:");
                console.log(`${I3.focusedMonitor.focusedWorkspace.name} and ${I3.focusedWorkspace.name}`);
            });
            I3.focusedMonitorChanged.connect(() => {
                console.log("Changed to monitor", I3.focusedMonitor.name);
            });
            for (const monitor of I3.monitors.values) {
                console.log("Monitor:", monitor.name);
                longLog && console.log(stringify(monitor));
            }
            for (const workspace of I3.workspaces.values) {
                console.log("Workspace:", workspace.name);
                longLog && console.log(stringify(workspace));
                I3.dispatch(`workspace ${workspace.num}`);
            }
            I3.dispatch(`workspace 13`);
            console.log(`Changing workspace response: ${stringify(I3.dispatch(`exec ${program}`))}`);
            console.log(`Bad request response: ${stringify(I3.dispatch(`quickshell`))}`);
            I3.dispatch(`workspace 1`);
            I3.refreshWorkspaces();
            I3.refreshMonitors();
            console.log(stringify(I3.monitorFor(panel.screen)));
        }
    }
}
```

</p>
</details>

**Comment by outfoxxed** [comment]
Until you commit the transaction you shouldn't access data that is part of it, as it might not have been read. Also, rolling back the transaction to read again when you hit something invalid will just hang. On my phone rn, will do a proper code review again later.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [reply src/x11/i3/ipc/connection.cpp]: Had to think about the expression for a minute for it to click but I see the intention now, mb.

File: src/x11/i3/ipc/connection.hpp
Comment (outfoxxed) [reply src/x11/i3/ipc/connection.hpp]: Sorry, haven't had time to go over this in detail before now, will comment on the larger issue in main review.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: This is a problem I missed the first time. Blocking on the main thread is almost never acceptable in quickshell, because not stalling the UI is extremely important. Especially important is to never block the main thread in a way that depends on an arbitrary response from another program. This is why the hyprland IPC makeRequest is asynchronous with a callback.

It does make things harder than they have to be, but I consider it very important to never leave the main thread blocked on IO.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: Similarly to the above, we can't return the response here without blocking. We need some kind of promise mechanism, but it's non-trivial to get right. I'd be happy to get input on it via the matrix channel or something if you have any, but don't try to deal with it in this PR.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp:208]: I was under the impression that workspace focus is per-monitor in i3, in which case this should be a function on I3Monitor instead. Does i3 allow moving a workspace to another monitor by setting it as the focused one?

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: given this, do we need refreshMonitors/refreshWorkspaces to handle responses, or can it just use the same handler as any other events?

Also please move it to a function to match the code style. I'll try to make that clearer in the future.

File: src/x11/i3/ipc/connection.hpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.hpp]: given that we can't make requests synchronously until promises are implemented, can we remove this?

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: `rawEvent` is a really weird name for this. Try `eventStream` or something that signifies its an input stream and not a singular event.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: Use std::array instead of a heap allocated, dynamically sized buffer.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: I think receiving a bad magic sequence isn't a recoverable error. I would tear down the connection and re-create it, or just print an error in the log and stop reading, because something is definitely wrong. Breaking at least will not recover this.

**Comment by outfoxxed** [comment]
While reviewing, I was going to test it and it crashed instantly under hyprland. It shouldn't work but it shouldn't crash either.
image

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: If you're going to log things, make it detailed enough to be useful.

File: src/x11/i3/ipc/monitor.hpp
Comment (outfoxxed) [inline src/x11/i3/ipc/monitor.hpp:68]: is there a reason we can't set these from qml via props or invocation?

**Review by outfoxxed — CHANGES_REQUESTED** [review]
Might've missed things again, but this is what I saw this time. Apologies for the repeated re-reviews but I want to be sure quickshell's integrations are stable.

File: src/x11/i3/ipc/connection.cpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.cpp]: Alright, means I will just drop the other socket and remove the return value

File: src/x11/i3/ipc/connection.cpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.cpp:208]: There can only ever be one focused workspace at all times (i.e only one workspace has focused = true), they do also return whether a workspace is visible or not, and if a workspace is visible, it is the one present in the current_workspace field you get from `get_outputs`

File: src/x11/i3/ipc/connection.cpp
Comment (nydragon) [reply src/x11/i3/ipc/connection.cpp]: Monitor events do not return any data, at least on Sway, there we'd need to fetch all outputs:

<details><summary>man</summary>
<p>

   0x80000001. OUTPUT
       Sent whenever an output is added, removed, or its configuration is changed.  The
       event  is a single object with the property change, which is a string containing
       the reason for the change. Currently, the only value for change is unspecified.

       Example Event:
           {
                "change": "unspecified"
           }


</p>
</details> 

But I am unsure about what exactly you mean by this message, events are generally handled in onEvent, and considering the previous comments, I will probably move the handling of get_outputs/get_workspaces to there and refresh* will only contain the request

File: src/x11/i3/ipc/monitor.hpp
Comment (nydragon) [reply src/x11/i3/ipc/monitor.hpp:68]: I did not intend to allow changing focused monitor that way, that's possible through the dispatch function for now, but it could definitely make sense

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [reply src/x11/i3/ipc/connection.cpp:208]: alright, not a problem.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [reply src/x11/i3/ipc/connection.cpp]: That sounds good to me.

File: src/x11/i3/ipc/monitor.hpp
Comment (outfoxxed) [reply src/x11/i3/ipc/monitor.hpp:68]: dw about it for now.

**Comment by nydragon** [comment]
Should be all fixed now

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: Don't need to reconnect for this one, as long as the event stream is still well-formed.

File: src/x11/i3/ipc/connection.cpp
Comment (outfoxxed) [inline src/x11/i3/ipc/connection.cpp]: ;

**Review by outfoxxed — CHANGES_REQUESTED** [review]
Didn't see anything else at a glance, will give it a more thorough lookover again shortly.

**Comment by outfoxxed** [comment]
We shouldn't be trying to write to the socket unless we're actually connected. I think hyprland IPC also does this though, now that I think about it. Will fix that later.
```
build/src/quickshell -p testi3.qml
  INFO: Launching config: "/home/admin/programming/outfoxxed/quickshell/i3/testi3.qml"
  INFO: Shell ID: "c4c77744f223921c7870881fbb46b0a8" Path ID "c4c77744f223921c7870881fbb46b0a8"
  INFO: Saving logs to "/run/user/1000/quickshell/by-id/jij3guvdns/log.log"
  WARN quickshell.I3.ipc: $I3SOCK is unset. Trying $SWAYSOCK.
  WARN quickshell.I3.ipc: $SWAYSOCK and I3SOCK are unset. Cannot connect to socket.
 DEBUG qml: Connecting to
  WARN: QIODevice::write (QLocalSocket): device not open
  WARN: QIODevice::write (QLocalSocket): device not open
 DEBUG qml: Changing workspace response: undefined
  WARN: QIODevice::write (QLocalSocket): device not open
 DEBUG qml: Bad request response: undefined
  WARN: QIODevice::write (QLocalSocket): device not open
  WARN: QIODevice::write (QLocalSocket): device not open
  WARN: QIODevice::write (QLocalSocket): device not open
 DEBUG qml: null
  INFO: Configuration Loaded
  ```

File: src/x11/i3/ipc/CMakeLists.txt
Comment (outfoxxed) [inline src/x11/i3/ipc/CMakeLists.txt]: indent is wrong

**Comment by outfoxxed** [comment]
Failing lints
```
/home/admin/programming/outfoxxed/quickshell/i3/src/x11/i3/ipc/qml.cpp:3:1: error: included header qjsonarray.h is not used directly [misc-include-cleaner,-warnings-as-errors]
    3 | #include <qjsonarray.h>
      | ^~~~~~~~~~~~~~~~~~~~~~~
    4 | #include <qobject.h>
```
```
/home/admin/programming/outfoxxed/quickshell/i3/src/x11/i3/ipc/connection.cpp:96:34: error: class member accesses must explicitly use thisptr [tidyfox-explicit-thisptr,-warnings-as-errors]
   96 |         for (auto& [type, data]: I3Ipc::parseResponse()) {
      |                                         ^
      |                                         this->
/home/admin/programming/outfoxxed/quickshell/i3/src/x11/i3/ipc/connection.cpp:135:7: error: no header providing "strncmp" is directly included [misc-include-cleaner,-warnings-as-errors]
    3 |                 if (strncmp(buffer.data(), MAGIC.data(), 6) != 0) {
      |                     ^
/home/admin/programming/outfoxxed/quickshell/i3/src/x11/i3/ipc/connection.cpp:377:13: error: method 'handleRunCommand' can be made static [readability-convert-member-functions-to-static,-warnings-as-errors]
  377 | void I3Ipc::handleRunCommand(I3IpcEvent* event) {
      |             ^
/home/admin/programming/outfoxxed/quickshell/i3/src/x11/i3/ipc/connection.cpp:380:3: error: variable 'success' of type 'bool' can be declared 'const' [misc-const-correctness,-warnings-as-errors]
  380 |                 bool success = obj["success"].toBool();
      |                 ^
      |                      const
/home/admin/programming/outfoxxed/quickshell/i3/src/x11/i3/ipc/connection.cpp:383:4: error: variable 'error' of type 'QString' can be declared 'const' [misc-const-correctness,-warnings-as-errors]
  383 |                         QString error = obj["error"].toString();
      |                         ^
      |                                 const
```

**Comment by nydragon** [comment]
Fixed

**Review by outfoxxed — APPROVED** [review]
Thanks!

**Pull request merged**
